src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt

```diff
@@ -292,27 +316,84 @@ abstract class AbstractCheckTask(
         if (prevState != State.PUBLISHED) {
             completeEvent(prevState == State.TIMEOUT)
             _endTime = Instant.now()
-            LOGGER.debug("Sending event tree id '{}' parent id '{}'", rootEvent.id, parentEventID)
-           val batch = EventBatch.newBuilder()
-               .setParentEventId(parentEventID)
-               .addAllEvents(rootEvent.toProtoEvents(parentEventID.id))
-               .build()
+
+            val events = rootEvent.toListProto(parentEventID)
+            val batches = batch(events, parentEventID)
 
             RESPONSE_EXECUTOR.execute {
-                try {
-                    eventBatchRouter.send(batch, "publish", "event")
-                    if (LOGGER.isDebugEnabled) {
-                        LOGGER.debug("Sent event batch '{}'", shortDebugString(batch))
+                batches.forEach { batch ->
+                    LOGGER.debug("Sending event batch parent id '{}'", parentEventID.id)
+                    try {
+                        eventBatchRouter.send(batch)
+                        if (LOGGER.isDebugEnabled) {
+                            LOGGER.debug("Sent event batch '{}'", shortDebugString(batch))
+                        }
+                    } catch (e: Exception) {
+                        LOGGER.error("Can not send event batch '{}'", shortDebugString(batch), e)
                     }
-                } catch (e: Exception) {
-                    LOGGER.error("Can not send event batch '{}'", shortDebugString(batch), e)
                 }
             }
         } else {
             LOGGER.debug("Event tree id '{}' parent id '{}' is already published", rootEvent.id, parentEventID)
         }
     }
 
+    private fun batch(events: List<ProtoEvent>, parentEventID: EventID): List<EventBatch> {
+        val result = ArrayList<EventBatch>()
+        val eventGroups = events.groupBy(ProtoEvent::getParentId)
+        batch(result, eventGroups, parentEventID)
+        return result
+    }
+
+    private fun batch(result: MutableList<EventBatch>, eventGroups: Map<EventID, List<ProtoEvent>>, eventID: EventID) {
+        val factory = { EventBatch.newBuilder().setParentEventId(eventID) }
+
+        var builder = factory.invoke()
+
+        eventGroups[eventID]?.forEach { event ->
+            val checkedEvent = checkAndRebuild(event)
+
+            LOGGER.debug("Process ${checkedEvent.name} ${checkedEvent.type}")
+            if(eventGroups.containsKey(checkedEvent.id)) {
+                builder.addEvents(checkedEvent)
+                result.add(checkAndBuild(builder))
+                builder = factory.invoke()
```

Comment:
Is it possible if we have several events with the same parent ID and one of them has child events and another one doesn't?
If it is possible then we can have the following situation:
1) We process the event without child events -> add it to the builder (line 370)
2) We process the event with child events -> add it to the builder (line 358) -> convert builder to the batch (batch has 2 events)

In that case, we have the event that has child events in the batch with several events. And it is not allowed If I remember correctly. Please, correct me if I am wrong

Reply:
Yes, you right. We should guarantee that event with children are published as a single event